share/man/man8/nanobsd.8
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2006 Daniel Gerzo <danger@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd April 17, 2026
.Dt NANOBSD 8
.Os
.Sh NAME
.Nm nanobsd.sh
.Nd create an embedded FreeBSD system image
.Sh SYNOPSIS
.Nm
.Op Fl BbfhIiKknPpqvWwX
.Op Fl c Ar config-file
.Sh DESCRIPTION
The
.Nm
utility is a script which produces a minimal implementation of
.Fx
(called
.Nm NanoBSD ) ,
which typically fits on a small media such as an SD card,
or other mass storage medium.
It can be used to build specialized install images, designed for easy
installation and maintenance.
.Pp
The following options are available:
.Bl -tag -width ".Fl c Ar config-file" -offset indent
.It Fl B
Skip the install stages (both for kernel and world).
.It Fl b
Skip the build stages (both for kernel and world).
.It Fl c Ar config-file
Specify the configuration file to use.
.It Fl f
Skip the code slice extraction.
.It Fl h
Display usage information.
.It Fl I
Build the disk image from an existing build/install.
.It Fl i
Skip the disk image build stage.
.It Fl K
Skip the
.Cm installkernel
stage of the build.
.It Fl k
Skip the
.Cm buildkernel
stage of the build.
.It Fl n
Do not cleanup before each build stage.
This suppresses the normal cleanup work done before the
.Cm buildworld
stage and adds -DNO_CLEAN to the make command line
used for each build stage (world and kernel).
.It Fl P
Use pre-compiled binaries instead of building from source.
By default, packages from
.Xr freebsd-base 7
are used.
When
.Va NANO_NOPKGBASE
is set, legacy distribution archive sets
.Pq such as Pa base.txz No and Pa kernel.txz
defined in
.Va NANO_DISTRIBUTIONS
are used instead.
See the
.Ev DISTRIBUTIONS
variable in
.Xr bsdinstall 8
for related distribution set names.
.It Fl p
Don't prepare the image.
Skip running of the customization and early customization scripts for
incremental image refinement from world, kernel, or packages.
.It Fl q
Make output more quiet.
.It Fl v
Make output more verbose.
.It Fl W
Skip the
.Cm installworld
stage of the build.
.It Fl w
Skip the
.Cm buildworld
stage of the build.
.It Fl X
Make
.Cm native-xtools .
.El
.Pp
The features of
.Nm NanoBSD
include:
.Pp
.Bl -bullet -offset indent -compact
.It
Ports and packages work as in
.Fx .
Every single application can be installed and used in a
.Nm NanoBSD
image, the same way as in
.Fx .
.It
No missing functionality.
If it is possible to do something with
.Fx ,
it is possible to do the same thing with
.Nm NanoBSD ,
unless the specific feature or features were explicitly removed from the
.Nm NanoBSD
image when it was created.
.It
Everything is read-only at run-time.
It is safe to pull the power-plug.
There is no necessity to run
.Xr fsck 8
after a non-graceful shutdown of the system.
.It
Easy to build and customize.
Making use of just one shell script and one configuration file it is
possible to build reduced and customized images satisfying any arbitrary
set of requirements.
.El
.Ss NanoBSD Media Layout
The mass storage medium is divided into three parts by default (which
are normally mounted read-only):
.Pp
.Bl -bullet -offset indent -compact
.It
Two image partitions:
.Li code#1
and
.Li code#2 .
.It
The configuration file partition, which can be mounted under the
.Pa /cfg
directory at run time.
.El
.Pp
The
.Pa /etc
and
.Pa /var
directories are
.Xr md 4
(malloc backed) disks.
.Pp
The configuration file partition persists under the
.Pa /cfg
directory.
It contains files for
.Pa /etc
directory and is briefly mounted read-only right after the system boot,
therefore it is required to copy modified files from
.Pa /etc
back to the
.Pa /cfg
directory if changes are expected to persist after the system restarts.
.Sh BUILDING Nm NanoBSD
A
.Nm NanoBSD
image is built using a simple
.Nm
shell script, which can be
found in the
.Pa src/tools/tools/nanobsd
directory.
This script creates a bootable image, which can be copied on the storage
medium using the
.Xr dd 1
utility.
.Pp
The necessary commands to build and install a
.Nm NanoBSD
image are:
.Bd -literal -offset indent
cd /usr/src/tools/tools/nanobsd
sh nanobsd.sh
cd /usr/obj/nanobsd.full
dd if=_.disk.full of=/dev/da0 bs=64k
.Ed
.Sh CUSTOMIZING Nm NanoBSD
This is probably the most important and most interesting feature of
.Nm NanoBSD .
This is also where you will be spending most of the time when developing with
.Nm NanoBSD .
.Pp
Customization is done in two ways:
.Pp
.Bl -bullet -offset indent -compact
.It
Configuration options.
.It
Custom functions.
.El
.Pp
With configuration settings, it is possible to configure options passed
to both the
.Cm buildworld
and
.Cm installworld
stages of the
.Nm NanoBSD
build process, as well as internal options passed to the main build
process of
.Nm NanoBSD .
Through these options it is possible to cut the system down, so it will
fit on as little as 64MB.
You can use the configuration options to trim down the system
even more, until it will consist of just the kernel and two or three
files in the userland.
.Pp
The configuration file consists of configuration options, which override
the default values.
The most important directives are:
.Bl -tag -width ".Va CONF_INSTALL" -offset indent
.It Va NANO_NAME
Build name (used to construct the working directory names).
.It Va NANO_SRC
Path to the source tree used to build the image.
.It Va NANO_KERNEL
Name of the kernel configuration file used to build the kernel.
.It Va NANO_ARCH
Machine processor architecture to build.
Defaults to output of
.Cm uname -p .
.It Va NANO_BOOT0CFG
Controls the options passed to
.Xr boot0cfg 8 ;
these dictate
.Nm boot0 Ns 's
behaviour.
.It Va NANO_BOOTLOADER
The
.Nm boot0
loader to use relative to the
.Va NANO_WORLDDIR
variable.
This defaults to
.Pa boot/boot0sio
and should be overridden to
.Pa boot/boot0
to provide a VGA
console.
.It Va CONF_BUILD
Options passed to the
.Cm buildworld
stage of the build.
.It Va CONF_INSTALL
Options passed to the
.Cm installworld
stage of the build.
.It Va CONF_WORLD
Options passed to both the
.Cm buildworld
and
.Cm installworld
stages of the build.
.It Va FlashDevice
Defines the type of media to use.
Check the
.Pa FlashDevice.sub
file for more details.
.El
.Pp
For more configuration options, please check the
.Nm
script.
.Pp
To build
.Nm NanoBSD
image using the
.Pa nanobsd.conf
configuration file, use the following command:
.Bd -literal -offset indent
sh nanobsd.sh -c nanobsd.conf
.Ed
.Pp
It is possible to fine-tune
.Nm NanoBSD
using shell functions in the configuration file.
The following example illustrates the basic model of custom functions:
.Bd -literal -offset indent
cust_foo () (
	echo "bar=topless" > \\
	     ${NANO_WORLDDIR}/etc/foo
)
customize_cmd cust_foo
.Ed
.Pp
There are a few pre-defined customization functions ready for use:
.Bl -tag -width ".Cm cust_allow_ssh_root" -offset indent
.It Cm cust_comconsole
Disables
.Xr getty 8
on the virtual
.Xr syscons 4
or
.Xr vt 4
terminals
.Pq Pa /dev/ttyv*
and enables the use of the first serial port as the system
console.
.It Cm cust_allow_ssh_root
Allow root to log in via
.Xr sshd 8 .
.It Cm cust_install_files
Installs files from the
.Pa nanobsd/Files
directory, which contains some useful scripts for system administration.
.El
.Sh FILES
.Bl -tag -width ".Pa src/tools/tools/nanobsd" -compact
.It Pa src/tools/tools/nanobsd
Base directory of the
.Nm NanoBSD
build script.
.El
.Sh EXAMPLES
Making persistent changes to
.Pa /etc/resolv.conf :
.Bd -literal -offset indent
vi /etc/resolv.conf
\&...
mount /cfg
cp /etc/resolv.conf /cfg
umount /cfg
.Ed
.Pp
A more useful example of a customization function is the following,
which changes the default size of the
.Pa /etc
directory from 5MB to 30MB:
.Bd -literal -offset indent
cust_etc_size () (
	cd ${NANO_WORLDDIR}/conf
	echo 30000 > default/etc/md_size
)
customize_cmd cust_etc_size
.Ed
.Sh SEE ALSO
.Xr make.conf 5 ,
.Xr freebsd-base 7 ,
.Xr boot 8 ,
.Xr boot0cfg 8 ,
.Xr bsdinstall 8
.Sh HISTORY
The
.Nm
utility first appeared in
.Fx 6.0 .
.Sh AUTHORS
.An -nosplit
.Nm NanoBSD
was developed by
.An Poul-Henning Kamp Aq Mt phk@FreeBSD.org .
This manual page was written by
.An Daniel Gerzo Aq Mt danger@FreeBSD.org .


tools/tools/nanobsd/nanobsd.sh
#!/bin/sh
#
# Copyright (c) 2005 Poul-Henning Kamp.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
#

set -e

nanobsd_sh=$(realpath $0)
topdir=$(dirname ${nanobsd_sh})
. "${topdir}/defaults.sh"
. "${topdir}/customizations.sh"
. "${topdir}/_xxx_includes.subr" # XXX ideally, this file should not exist

is_defined() {
	case $(type $1 2>/dev/null) in
	*function) return 0 ;;
	*) return 1 ;;
	esac
}

legacy() {
	# Pull in legacy stuff on demand
	. "${topdir}/legacy.sh"
	NANO_PLAN="legacy"
}

default() {
	if [ "$NANO_FILE_SYSTEM" = "ZFS" ]; then
		. "${topdir}/zfs.sh"
	else
		. "${topdir}/ufs.sh"
	fi
	NANO_PLAN="${NANO_PLAN:-default}"
}

#######################################################################
# Parse arguments

do_clean=true
do_root=true
do_kernel=true
do_installkernel=true
do_world=true
do_code=true
do_installworld=true
do_image=true
do_native_xtools=false
do_precompiled=false
do_prep_image=true

set +e
args=$(getopt BKXWbc:fhiIknPpqUvw $*)
if [ $? -ne 0 ]; then
	usage
	exit 2
fi
set -e

set -- $args
for i
do
	case "$i"
	in
	-B)
		do_installworld=false
		do_installkernel=false
		shift
		;;
	-I)
		do_world=false
		do_kernel=false
		do_installworld=false
		do_installkernel=false
		do_prep_image=false
		do_image=true
		shift
		;;
	-K)
		do_installkernel=false
		shift
		;;
	-X)
		do_native_xtools=true
		shift
		;;
	-W)
		do_installworld=false
		shift
		;;
	-b)
		do_world=false
		do_kernel=false
		shift
		;;
	-c)
		# Make config file path available to the config file
		# itself so that it can access additional files relative
		# to its own location.
		NANO_CONFIG=$2
		. "$2"
		shift
		shift
		;;
	-f)
		do_code=false
		do_image=false
		shift
		;;
	-h)
		usage
		;;
	-i)
		do_image=false
		shift
		;;
	-k)
		do_kernel=false
		shift
		;;
	-n)
		do_clean=false
		shift
		;;
	-P)
		do_world=false
		do_kernel=false
		do_precompiled=true
		NANO_TIMESTAMP=
		shift
		;;
	-p)
		do_prep_image=false
		shift
		;;
	-q)
		PPLEVEL=$(($PPLEVEL - 1))
		shift
		;;
	-v)
		PPLEVEL=$(($PPLEVEL + 1))
		shift
		;;
	-U)
		do_root=false
		NANO_NOPRIV_BUILD=true
		shift
		;;
	-w)
		do_world=false
		shift
		;;
	--)
		shift
		break
	esac
done

if [ $# -gt 0 ]; then
	echo "$0: Extraneous arguments supplied"
	usage
fi

if $do_precompiled && [ "${NANO_PLAN:-legacy}" != "default" ]; then
	echo "$0: -P requires a config file that selects the 'default' plan" 1>&2
	exit 1
fi

# Transition hack -- If you get this warning, add 'legacy' to your nano config file
if [ -z "$NANO_PLAN" ]; then
	echo "Warning: no plan defined, assuming legacy config."
	legacy
fi

#######################################################################
# And then it is as simple as that...

# File descriptor 3 is used for logging output, see pprint
exec 3>&1
set_defaults_and_export

if [ ! -d "${NANO_TOOLS}" ]; then
	echo "NANO_TOOLS directory does not exist" 1>&2
	exit 1
fi

pprint 1 "NanoBSD image ${NANO_NAME} build starting"

if $do_prep_image; then
	run_early_customize
else
	pprint 2 "Skipping early customization for image prep (as instructed)"
fi

if $do_world; then
	if $do_clean; then
		clean_build
	else
		pprint 2 "Using existing build tree (as instructed)"
	fi
	make_conf_build
	build_world
else
	pprint 2 "Skipping buildworld (as instructed)"
fi

if $do_kernel; then
	if ! $do_world; then
		make_conf_build
	fi
	build_kernel
else
	pprint 2 "Skipping buildkernel (as instructed)"
fi

if [ -z "$NANO_NOPKGBASE" ]; then
	$do_precompiled || build_packages
	nano_setup_pkg_repo
	if [ -n "${NANO_METALOG}" ]; then
		nano_pkgbase_reset_metalog
	fi
else
	nano_fetch_distsets
	if [ -n "${NANO_METALOG}" ]; then
		nano_distset_metalog
	fi
fi

if $do_installworld; then
	clean_world
	if $do_precompiled || [ -z "$NANO_NOPKGBASE" ]; then
		install_world_binaries
	else
		make_conf_install
		install_world
	fi
	install_etc
else
	pprint 2 "Skipping installworld (as instructed)"
fi

if ${do_native_xtools}; then
	native_xtools
fi
if ${do_prep_image}; then
	setup_nanobsd_etc
fi
if $do_installkernel; then
	if $do_precompiled || [ -z "$NANO_NOPKGBASE" ]; then
		install_kernel_binaries
	else
		install_kernel
	fi
else
	pprint 2 "Skipping installkernel (as instructed)"
fi

if $do_precompiled && [ -n "$NANO_NOPKGBASE" ] && \
    ($do_installworld || $do_installkernel); then
	patch_precompiled
fi

if $do_prep_image; then
	run_customize
	setup_nanobsd
	prune_usr
	run_late_customize
	fixup_before_diskimage
else
	pprint 2 "Skipping image prep (as instructed)"
fi
if $do_code; then
	calculate_partitioning
	create_code_partition
	if $do_image; then
		if [ -z "${NANO_NOPRIV_BUILD}" ]; then
			create_diskimage
		else
			_create_diskimage
		fi
	else
		pprint 2 "Skipping image build (as instructed)"
	fi
else
	pprint 2 "Skipping code and image build (as instructed)"
fi
last_orders

pprint 1 "NanoBSD image ${NANO_NAME} completed"
